File: go/benchmark_manual.go
package main

import (
	"crypto/rand"
	"fmt"
	"math"
	"math/big"
	"math/rand"
	"time"
)

// runBenchmark executa múltiplas medições e calcula estatísticas
func runBenchmark(operation func(), iterations int, name string) map[string]float64 {
	var times []float64
	
	fmt.Printf("Executando benchmark: %s\n", name)
	
	for i := 0; i < iterations; i++ {
		start := time.Now()
		operation()
		duration := time.Since(start)
		times = append(times, float64(duration.Nanoseconds()))
		
		if (i+1)%10 == 0 {
			fmt.Printf("  Iteração %d/%d\n", i+1, iterations)
		}
	}
	
	// Calcular estatísticas
	var total float64
	for _, t := range times {
		total += t
	}
	mean := total / float64(len(times))
	
	var variance float64
	for _, t := range times {
		variance += (t - mean) * (t - mean)
	}
	variance /= float64(len(times))
	stdDev := math.Sqrt(variance)
	
	min := times[0]
	max := times[0]
	for _, t := range times {
		if t < min {
			min = t
		}
		if t > max {
			max = t
		}
	}
	
	stats := make(map[string]float64)
	stats["mean_ns"] = mean
	stats["std_dev_ns"] = stdDev
	stats["min_ns"] = min
	stats["max_ns"] = max
	stats["total_ns"] = total
	
	return stats
}

// printStats imprime estatísticas
func printStats(stats map[string]float64, operation string) {
	fmt.Printf("\n=== Estatísticas para %s ===\n", operation)
	fmt.Printf("Média: %.2f ns\n", stats["mean_ns"])
	fmt.Printf("Desvio Padrão: %.2f ns\n", stats["std_dev_ns"])
	fmt.Printf("Mínimo: %.2f ns\n", stats["min_ns"])
	fmt.Printf("Máximo: %.2f ns\n", stats["max_ns"])
	fmt.Printf("Total: %.2f ns\n", stats["total_ns"])
}

// benchmarkManualRSA executa benchmark da implementação manual
func benchmarkManualRSA() {
	fmt.Println("=== Benchmark Manual RSA ===")
	fmt.Println("Usando funções nativas de tempo\n")
	
	iterations := 100
	
	// Benchmark de geração de chaves
	keyGenStats := runBenchmark(
		func() {
			_ = NewRSAKey(2048)
		},
		iterations,
		"Geração de Chaves",
	)
	printStats(keyGenStats, "Geração de Chaves")
	
	// Gerar uma chave para os testes de criptografia
	key := NewRSAKey(2048)
	testMessage := big.NewInt(12345)
	
	// Benchmark de criptografia
	encryptionStats := runBenchmark(
		func() {
			_ = Encrypt(testMessage, key)
		},
		iterations,
		"Criptografia",
	)
	printStats(encryptionStats, "Criptografia")
	
	// Criptografar uma mensagem para o teste de descriptografia
	ciphertext := Encrypt(testMessage, key)
	
	// Benchmark de descriptografia
	decryptionStats := runBenchmark(
		func() {
			_ = Decrypt(ciphertext, key)
		},
		iterations,
		"Descriptografia",
	)
	printStats(decryptionStats, "Descriptografia")
	
	// Comparação de performance
	fmt.Println("\n=== Comparação de Performance ===")
	fmt.Printf("Geração de Chaves: %.2f ns (média)\n", keyGenStats["mean_ns"])
	fmt.Printf("Criptografia: %.2f ns (média)\n", encryptionStats["mean_ns"])
	fmt.Printf("Descriptografia: %.2f ns (média)\n", decryptionStats["mean_ns"])
	
	// Verificar se a descriptografia funciona corretamente
	decrypted := Decrypt(ciphertext, key)
	if testMessage.Cmp(decrypted) == 0 {
		fmt.Println("✓ Verificação de integridade: OK")
	} else {
		fmt.Println("✗ Verificação de integridade: FALHOU")
	}
}

// benchmarkKeySizes executa benchmark de diferentes tamanhos de chave
func benchmarkKeySizes() {
	fmt.Println("\n=== Benchmark de Diferentes Tamanhos de Chave ===")
	
	keySizes := []int{512, 1024, 2048}
	iterations := 50
	
	for _, bits := range keySizes {
		fmt.Printf("\n--- Chave de %d bits ---\n", bits)
		
		stats := runBenchmark(
			func() {
				_ = NewRSAKey(bits)
			},
			iterations,
			fmt.Sprintf("Geração de Chave %d bits", bits),
		)
		
		printStats(stats, fmt.Sprintf("Geração de Chave %d bits", bits))
	}
}

// benchmarkMessageSizes executa benchmark de diferentes tamanhos de mensagem
func benchmarkMessageSizes() {
	fmt.Println("\n=== Benchmark de Diferentes Tamanhos de Mensagem ===")
	
	key := NewRSAKey(2048)
	messageSizes := []int{100, 1000, 10000, 100000}
	iterations := 50
	
	for _, size := range messageSizes {
		fmt.Printf("\n--- Mensagem de %d bytes ---\n", size)
		
		testMessage := big.NewInt(int64(size))
		
		encryptionStats := runBenchmark(
			func() {
				_ = Encrypt(testMessage, key)
			},
			iterations,
			fmt.Sprintf("Criptografia %d bytes", size),
		)
		
		printStats(encryptionStats, fmt.Sprintf("Criptografia %d bytes", size))
	}
}

func main() {
	benchmarkManualRSA()
	benchmarkKeySizes()
	benchmarkMessageSizes()
} 

File: rust/benches/benchmark_manual.rs
use std::time::{Duration, Instant};
use std::collections::HashMap;
use rsa_manual::{RSAKey, encrypt, decrypt, generate_rsa_keys};

// Função para executar múltiplas medições e calcular estatísticas
fn run_benchmark<F>(func: F, iterations: usize, name: &str) -> HashMap<String, f64>
where
    F: Fn() -> Duration,
{
    let mut times = Vec::new();
    
    println!("Executando benchmark: {}", name);
    
    for i in 0..iterations {
        let duration = func();
        times.push(duration.as_nanos() as f64);
        
        if (i + 1) % 10 == 0 {
            println!("  Iteração {}/{}", i + 1, iterations);
        }
    }
    
    // Calcular estatísticas
    let total: f64 = times.iter().sum();
    let mean = total / times.len() as f64;
    
    let variance: f64 = times.iter()
        .map(|&x| (x - mean).powi(2))
        .sum::<f64>() / times.len() as f64;
    let std_dev = variance.sqrt();
    
    let min = times.iter().fold(f64::INFINITY, |a, &b| a.min(b));
    let max = times.iter().fold(f64::NEG_INFINITY, |a, &b| a.max(b));
    
    let mut stats = HashMap::new();
    stats.insert("mean_ns".to_string(), mean);
    stats.insert("std_dev_ns".to_string(), std_dev);
    stats.insert("min_ns".to_string(), min);
    stats.insert("max_ns".to_string(), max);
    stats.insert("total_ns".to_string(), total);
    
    stats
}

// Função para imprimir estatísticas
fn print_stats(stats: &HashMap<String, f64>, operation: &str) {
    println!("\n=== Estatísticas para {} ===", operation);
    println!("Média: {:.2} ns", stats["mean_ns"]);
    println!("Desvio Padrão: {:.2} ns", stats["std_dev_ns"]);
    println!("Mínimo: {:.2} ns", stats["min_ns"]);
    println!("Máximo: {:.2} ns", stats["max_ns"]);
    println!("Total: {:.2} ns", stats["total_ns"]);
}

// Função principal do benchmark manual
pub fn main() {
    println!("=== Benchmark Manual RSA ===");
    println!("Usando funções nativas de tempo\n");
    
    let iterations = 100;
    
    // Benchmark de geração de chaves
    let key_gen_stats = run_benchmark(
        || {
            let start = Instant::now();
            let _key = RSAKey::new();
            start.elapsed()
        },
        iterations,
        "Geração de Chaves"
    );
    print_stats(&key_gen_stats, "Geração de Chaves");
    
    // Gerar uma chave para os testes de criptografia
    let key = RSAKey::new();
    let test_message = 12345u128;
    
    // Benchmark de criptografia
    let encryption_stats = run_benchmark(
        || {
            let start = Instant::now();
            let _ciphertext = encrypt(test_message, &key);
            start.elapsed()
        },
        iterations,
        "Criptografia"
    );
    print_stats(&encryption_stats, "Criptografia");
    
    // Criptografar uma mensagem para o teste de descriptografia
    let ciphertext = encrypt(test_message, &key);
    
    // Benchmark de descriptografia
    let decryption_stats = run_benchmark(
        || {
            let start = Instant::now();
            let _plaintext = decrypt(ciphertext, &key);
            start.elapsed()
        },
        iterations,
        "Descriptografia"
    );
    print_stats(&decryption_stats, "Descriptografia");
    
    // Comparação de performance
    println!("\n=== Comparação de Performance ===");
    println!("Geração de Chaves: {:.2} ns (média)", key_gen_stats["mean_ns"]);
    println!("Criptografia: {:.2} ns (média)", encryption_stats["mean_ns"]);
    println!("Descriptografia: {:.2} ns (média)", decryption_stats["mean_ns"]);
    
    // Verificar se a descriptografia funciona corretamente
    let decrypted = decrypt(ciphertext, &key);
    if test_message == decrypted {
        println!("✓ Verificação de integridade: OK");
    } else {
        println!("✗ Verificação de integridade: FALHOU");
    }
}

// Função para benchmark de diferentes tamanhos de chave
pub fn benchmark_key_sizes() {
    println!("\n=== Benchmark de Diferentes Tamanhos de Chave ===");
    
    let key_sizes = vec![512, 1024, 2048];
    let iterations = 50;
    
    for &bits in &key_sizes {
        println!("\n--- Chave de {} bits ---", bits);
        
        let stats = run_benchmark(
            || {
                let start = Instant::now();
                let _keys = generate_rsa_keys(bits);
                start.elapsed()
            },
            iterations,
            &format!("Geração de Chave {} bits", bits)
        );
        
        print_stats(&stats, &format!("Geração de Chave {} bits", bits));
    }
}

// Função para benchmark de diferentes tamanhos de mensagem
pub fn benchmark_message_sizes() {
    println!("\n=== Benchmark de Diferentes Tamanhos de Mensagem ===");
    
    let key = RSAKey::new();
    let message_sizes = vec![100, 1000, 10000, 100000];
    let iterations = 50;
    
    for &size in &message_sizes {
        println!("\n--- Mensagem de {} bytes ---", size);
        
        let test_message = size as u128;
        
        let encryption_stats = run_benchmark(
            || {
                let start = Instant::now();
                let _ciphertext = encrypt(test_message, &key);
                start.elapsed()
            },
            iterations,
            &format!("Criptografia {} bytes", size)
        );
        
        print_stats(&encryption_stats, &format!("Criptografia {} bytes", size));
    }
} 